phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java

```diff
@@ -556,35 +564,55 @@ private static String toString(List<byte[]> gps) {
-                }
-                try {
-                    while (guideIndex < gpsSize && (currentGuidePost.compareTo(endKey) <= 0 || endKey.length == 0)) {
-                        Scan newScan = scanRanges.intersectScan(scan, currentKeyBytes, currentGuidePostBytes, keyOffset,
-                                false);
-                        estimatedRows += gps.getRowCounts().get(guideIndex);
-                        estimatedSize += gps.getByteCounts().get(guideIndex);
-                        scans = addNewScan(parallelScans, scans, newScan, currentGuidePostBytes, false, regionLocation);
-                        currentKeyBytes = currentGuidePost.copyBytes();
-                        currentGuidePost = PrefixByteCodec.decode(decoder, input);
-                        currentGuidePostBytes = currentGuidePost.copyBytes();
-                        guideIndex++;
-                    }
-                } catch (EOFException e) {}
-                Scan newScan = scanRanges.intersectScan(scan, currentKeyBytes, endKey, keyOffset, true);
-                if (isLocalIndex) {
-                    if (newScan != null) {
-                        newScan.setAttribute(EXPECTED_UPPER_REGION_KEY, endRegionKey);
-                    } else if (!scans.isEmpty()) {
-                        scans.get(scans.size()-1).setAttribute(EXPECTED_UPPER_REGION_KEY, endRegionKey);
-                    }
-                }
-                scans = addNewScan(parallelScans, scans, newScan, endKey, true, regionLocation);
-                currentKeyBytes = endKey;
+                if (Bytes.compareTo(scan.getStartRow(), context.getScan().getStartRow()) != 0
+                     || Bytes.compareTo(scan.getStopRow(), context.getScan().getStopRow()) != 0) {
+                    Scan newScan = ScanUtil.newScan(scan);
+                    if(ScanUtil.isLocalIndex(scan)) {
+                        newScan.setStartRow(regionInfo.getStartKey());
+                        newScan.setAttribute(SCAN_ACTUAL_START_ROW, regionInfo.getStartKey());
+                        newScan.setStopRow(regionInfo.getEndKey());
+                        newScan.setAttribute(EXPECTED_UPPER_REGION_KEY, regionInfo.getEndKey());
+                    } else {
+                        if(Bytes.compareTo(scan.getStartRow(), regionInfo.getStartKey())<=0) {
+                            newScan.setAttribute(SCAN_ACTUAL_START_ROW, regionInfo.getStartKey());
+                            newScan.setStartRow(regionInfo.getStartKey());
+                        }
+                        if(scan.getStopRow().length == 0 || (regionInfo.getEndKey().length != 0 && Bytes.compareTo(scan.getStopRow(), regionInfo.getEndKey())>0)) {
+                            newScan.setStopRow(regionInfo.getEndKey());
+                        }
+                    }   
+                    scans = addNewScan(parallelScans, scans, newScan, endKey, true, regionLocation);
+                 } else {
```

Comment:
Can we remove the if (isLocalIndex) checks below since the above if statement would have been entered instead?